src/snarky-typescript/src/index.ts

```diff
@@ -0,0 +1,517 @@
+}
+
+interface CircuitVarCtor<F, Circuit, Var extends CircuitVar<F, Circuit>> {
+  neg(x: FieldConvertible): Var;
+  neg(x: Var | F): Var;
+
+  inv(x: FieldConvertible): Var;
+  inv(x: Var | F): Var;
+
+  add(x: FieldConvertible, y: FieldConvertible): Var;
+  add(x: FieldConvertible, y: Var | F): Var;
+  add(x: Var | F, y: FieldConvertible): Var;
+  add(x: Var | F, y: Var | F): Var;
+
+  sub(x: FieldConvertible, y: FieldConvertible): Var;
+  sub(x: FieldConvertible, y: F): Var;
+  sub(x: FieldConvertible, y: Var): Var;
+  sub(x: Var | F, y: FieldConvertible): Var;
+  sub(x: Var | F, y: F): Var;
+  sub(x: Var | F, y: Var): Var;
+
+  mul(x: FieldConvertible, y: FieldConvertible): Var;
+  mul(x: FieldConvertible, y: Var | F): Var;
+  mul(x: Var | F, y: FieldConvertible): Var;
+  mul(x: Var | F, y: Var | F): Var;
+
+  div(x: FieldConvertible, y: FieldConvertible): Var;
+  div(x: FieldConvertible, y: Var | F): Var;
+  div(x: Var | F, y: FieldConvertible): Var;
+  div(x: Var | F, y: Var | F): Var;
+
+  scale(x: Var | F, scalar: FieldConvertible): Var;
+  scale(x: Var | F, scalar: F): Var;
+
+  equals(x: FieldConvertible, y: FieldConvertible): Var;
+  equals(x: FieldConvertible, y: Var | F): Var;
+  equals(x: Var | F, y: FieldConvertible): Var;
+  equals(x: Var | F, y: Var | F): Var;
+
+  assertBoolean(x: FieldConvertible): BoolVar<F, Circuit, Var>;
+  assertBoolean(x: Var | F): BoolVar<F, Circuit, Var>;
+  assertBoolean(x: Var): BoolVar<F, Circuit, Var>;
+
+  assertEquals(x: FieldConvertible, y: FieldConvertible): void;
+  assertEquals(x: FieldConvertible, y: Var | F): void;
+  assertEquals(x: Var | F, y: FieldConvertible): void;
+  assertEquals(x: Var | F, y: Var | F): void;
+
+  toBits(x: FieldConvertible): BoolVar<F, Circuit, Var>[];
+  toBits(x: Var | F): BoolVar<F, Circuit, Var>[];
+
+  ofBits(x: (boolean | BoolVar<F, Circuit, Var>)[]): Var;
+
+  toFieldElements(x: FieldConvertible): [Var];
+  toFieldElements(x: Var | F): [Var];
+
+  ofFieldElements(x: (FieldConvertible | F | Var)[]): Var;
+
+  store(x: Var, val: FieldConvertible): void;
+  store(x: Var, val: F): void;
+  store(x: Var, f: () => F): void;
+  store(x: Var, f: () => FieldConvertible): void;
+
+  read(x: FieldConvertible): F;
+  read(x: F): F;
+  read(x: Var): F;
+
+  new (): Var;
+  new (options: {isPublic: boolean}): Var;
+
+  new (x: FieldConvertible): Var;
+  new (x: F): Var;
+  new (x: Var): Var;
+  new (f: () => FieldConvertible | F | Var): Var;
+
+  new (x: FieldConvertible, options: {isPublic: boolean}): Var;
+  new (x: F, options: {isPublic: boolean}): Var;
+  new (x: Var, options: {isPublic: boolean}): Var;
+  new (f: () => F, options: {isPublic: boolean}): Var;
+  new (f: () => FieldConvertible, options: {isPublic: boolean}): Var;
+
+  new (x: FieldConvertible, isPublic: boolean): Var;
+  new (x: F, isPublic: boolean): Var;
+  new (x: Var, isPublic: boolean): Var;
+  new (f: () => F, isPublic: boolean): Var;
+  new (f: () => FieldConvertible, isPublic: boolean): Var;
```

Comment:
This constructor type is very overloaded. Is there a semantic difference here between constructing a `CircuitVar` using a value vs a function that returns a value? Also, why do the constructors with additional options that accept functions not support `Var` as a return value, but the constructor without options does? (specifically, `new (f: () => FieldConvertible | F | Var): Var;`).

Reply:
Brandon suggested that these should be split into a separate `PublicInput` constructor, which would tidy up quite a few of the options.
For the functional argument, it's possible I was just bad with my testing, but the typechecker was giving errors when I had 2 different functional types. I'll try and see if there's a nicer way to do this.

Izzy has also proposed a `witness` function and then merging the two different kinds of field element (`FieldElement` and `CircuitVar`), which will let us get rid of some of these too.